docs/building-rpms-from-source.md

```diff
@@ -0,0 +1,166 @@
+# Building RPMs from source with mkosi
+
+If you want to build an RPM from source and install it within a mkosi
+image, you can do that with mkosi itself without using `mock`. The steps
+required are as follows:
+
+1. Install `Requires` dependencies in the image
+2. Install `BuildRequires` dependencies in the build overlay
+3. Install dynamic `BuildRequires` dependencies in the build overlay
+4. Build the RPM with `rpmbuild`
+5. Install the built rpms in the image
+
+In the following examples, we'll use mkosi itself and its Fedora RPM
+spec as an example.
+
+To keep things snappy, we execute the first 3 steps in a prepare
+script so that they're cached on subsequent runs of mkosi if the
+`Incremental=` setting is enabled.
+
+First, we need access to the upstream sources and the RPM spec and
+related files. These can be mounted into the current working directory
+when running mkosi scripts by using the `BuildSources=` setting. For
+example, in `mkosi.local.conf`, we could have the following settings:
+
+`mkosi.local.conf`:
+
+```conf
+[Content]
+BuildSources=../mkosi:mkosi
+             ../fedora/mkosi:mkosi/rpm
+```
+
+Which instructs mkosi to mount the local version of the mkosi upstream
+repository at `../mkosi` to `mkosi` in the current working directory
+when running mkosi. The Fedora RPM spec is mounted at `mkosi/rpm`.
+
+We use `rpmspec` and `rpmbuild` but these do not really support running
+from outside of the image that the RPM is being built in, so we have to
+make sure they're available inside the image:
+
+`mkosi.conf`:
+
+```conf
+[Content]
+Packages=rpm-build
+# If you don't want rpm-build in the final image.
+RemovePackages=rpm-build
+```
+
+The prepare script then looks as follows:
+
+`mkosi.prepare`:
+
+```shell
+#!/bin/sh
+set -e
+
+if [ "$1" = "build" ]; then
+    DEPS="--buildrequires"
+else
+    DEPS="--requires"
+fi
+
+mkosi-chroot \
+    rpmspec \
+    --query \
+    "$DEPS" \
+    --define "_topdir $CHROOT_SRCDIR/mkosi" \
+    --define "_sourcedir $CHROOT_SRCDIR/mkosi/rpm" \
+    "$CHROOT_SRCDIR/mkosi/rpm/mkosi.spec" \
+    | grep -E -v "mkosi" \
+    | xargs -d '\n' dnf install --best
+
+if [ "$1" = "build" ]; then
+    until mkosi-chroot \
+        sh -c 'cd $CHROOT_SRCDIR/mkosi && exec $0 "$@"' \
+        rpmbuild \
+        -bd \
+        --build-in-place \
+        --define "_topdir $CHROOT_SRCDIR/mkosi" \
+        --define "_sourcedir $CHROOT_SRCDIR/mkosi/rpm" \
+        --define "_build_name_fmt %%{NAME}-%%{VERSION}-%%{RELEASE}.%%{ARCH}.rpm" \
+        "$CHROOT_SRCDIR/mkosi/rpm/mkosi.spec"
+    do
+        if [ $? -ne 11 ]; then
+            exit $?
```

Comment:
Is this the same exit code, though? In POSIX shell `[` is a command and will have its own exit code, so for comparisons it's safer to safe it into a variable for reuse.